lotus-soup/testkit/role_miner.go

```diff
@@ -202,10 +218,17 @@ func PrepareMiner(t *TestEnvironment) (*LotusMiner, error) {
 		withMinerListenAddress(minerIP),
 	}
 
-	if t.StringParam("mining_mode") != "natural" {
+	if miningMode != "natural" {
 		mineBlock := make(chan func(bool, error))
 		minerOpts = append(minerOpts,
-			node.Override(new(*miner.Miner), miner.NewTestMiner(mineBlock, minerAddr)))
+			node.Override(new(*miner.Miner), miner.NewTestMiner(mineBlock, minerAddr)),
+		)
+
+		// replace the clock, setting it to the genesis timestamp + 1 epoch
+		// we are now ready to mine (and receive) the first block.
+		m.MockClock = clock.NewMock()
+		m.MockClock.Set(time.Unix(int64(genesisMsg.TimestampUnix+build.BlockDelaySecs), 0))
```

Comment:
Since we increase the clock upfront at the beginning of the `for` in `launchMinerCoordination()` consider not adding this initial `BlockDelaySecs` here (the miners complain they are mining in the past and receiving blocks with large delays). Still, we could consider a very slight time shift here to avoid the edge case of comparing _exactly_ equal times (which many parts of the Lotus code might not be prepared for).

Reply:
Yep, good recommendation. I fixed this in my overhaul, which I just pushed to this branch.